src/ui/tabbareventfilter.cc

```diff
@@ -0,0 +1,194 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "ui/tabbareventfilter.h"
+
+#include <QApplication>
+#include "ui/dockwidget_native.h"
+
+#include "ui/mainwindowwithdetachabledockwidgets.h"
+
+namespace veles {
+namespace ui {
+
+/*****************************************************************************/
+/* TabBarEventFilter */
+/*****************************************************************************/
+
+TabBarEventFilter::TabBarEventFilter(QObject* parent)
+    : QObject(parent), drag_init_pos_(0, 0) {}
+
+void TabBarEventFilter::tabMoved(int /*from*/, int /*to*/) {
+  if (dragged_tab_bar_ != nullptr) {
+    dragged_tab_index_ = dragged_tab_bar_->currentIndex();
+  }
+}
+
+void TabBarEventFilter::currentChanged(int index) {
+  auto main_window =
+      MainWindowWithDetachableDockWidgets::getParentMainWindow(sender());
+  if (main_window != nullptr) {
+    auto dock_widget = dynamic_cast<DockWidget*>(
+        main_window->tabToDockWidget(dynamic_cast<QTabBar*>(sender()), index));
+
+    if (dock_widget != nullptr) {
+      MainWindowWithDetachableDockWidgets::setActiveDockWidget(dock_widget);
+    }
+  }
+}
+
+bool TabBarEventFilter::eventFilter(QObject* watched, QEvent* event) {
+  auto* tab_bar = dynamic_cast<QTabBar*>(watched);
+  if (tab_bar == nullptr) {
+    return false;
+  }
+
+  connect(tab_bar, &QTabBar::currentChanged, this,
+          &TabBarEventFilter::currentChanged, Qt::UniqueConnection);
+
+  auto main_window =
+      MainWindowWithDetachableDockWidgets::getParentMainWindow(watched);
+
+  if (main_window != nullptr && !main_window->dockWidgetsWithNoTitleBars()) {
+    return false;
+  }
+
+  connect(tab_bar, &QTabBar::tabMoved, this, &TabBarEventFilter::tabMoved,
+          Qt::UniqueConnection);
+
+  if (event->type() != QEvent::MouseMove &&
+      event->type() != QEvent::MouseButtonPress &&
+      event->type() != QEvent::MouseButtonRelease &&
+      event->type() != QEvent::MouseButtonDblClick) {
+    return false;
+  }
+
+  auto* mouse_event = static_cast<QMouseEvent*>(event);
+
+  if (mouse_event->type() == QEvent::MouseMove) {
+    return mouseMove(tab_bar, mouse_event);
+  }
+  if (mouse_event->type() == QEvent::MouseButtonPress) {
+    return mouseButtonPress(tab_bar, mouse_event);
+  }
+  if (mouse_event->type() == QEvent::MouseButtonRelease) {
+    return mouseButtonRelease(tab_bar, mouse_event);
+  }
+  if (mouse_event->type() == QEvent::MouseButtonDblClick) {
+    return mouseButtonDblClick(tab_bar, mouse_event);
+  }
+  return false;
+}
+
+bool TabBarEventFilter::mouseMove(QTabBar* tab_bar, QMouseEvent* event) {
+  if (dragged_tab_bar_ != nullptr) {
+    bool horizontal_tabs =
+        dragged_tab_bar_->shape() == QTabBar::RoundedNorth ||
+            dragged_tab_bar_->shape() == QTabBar::RoundedSouth ||
```

Comment:
Are you sure that the indentation here is ok? It wasn't like that before you moved the class here.